### PR TITLE
feat: show remaining credits above the chat composer

### LINKED ISSUE #1063: Chat users cannot see credits or usage left anywhere; add tenant-scoped balance endpoint and chat visibility
## Problem

A signed-in chat user has no way to see credits or usage anywhere in the product. The only surface showing "usage left" today is the developer console (`apps/web-console/app/console/billing/page.tsx`), which lives in a different identity space than chat. The owner wants remaining credits visible in the chat interface itself.

## Findings (traced 2026-08-23, read-only)

**Data layer**
- Balance is DERIVED, never stored: `credit_ledger_entries` summed by `ledger.pgxRepository.GetBalance` (`apps/control-plane/internal/ledger/repository.go:142`), `available = posted - reserved`, with per-reservation clamping added by #918. Measured 6.2ms at 5000 ledger rows / 1667 reservations. Fine for page-load polling, not per-keystroke.
- `tenant_billing_accounts` bridges tenant id space (`tenant_users`) to account id space (`account_memberships`), one-to-one, service-role only, RLS enabled with zero policies BY DESIGN. No browser may ever learn this link.

**API layer**
- `GET /api/v1/accounts/current/credits/balance` exists (`internal/ledger/http.go:41`) returning `posted_credits / reserved_credits / available_credits`, but it authenticates through `EnsureViewerContext` over `account_memberships`. Chat sessions authenticate as OWUI users in `tenant_users`, a disjoint id space (`platform/role.go:65`). So the endpoint is unusable from chat as-is.

**UI layer**
- Chat shell (`vendor/open-webui/src/lib/hive/`) has zero billing references. Upstream admin analytics views are unrelated to personal remaining credits.

## Proposed shape (smallest correct change)

1. New control-plane internal route, e.g. `GET /internal/tenants/current/credits/balance`, authenticated by tenant identity, resolving `tenant_id -> tenant_billing_accounts.account_id -> ledger.Service.GetBalance`. Reuse GetBalance untouched.
2. OWUI Python backend calls it server-side and serves a trimmed `{available_credits}` to its own frontend under the existing OWUI session. No new browser-reachable cross-id-space surface.
3. Widget placement follows the settings IA spec (spec-2026-08-23-settings-ia.md); poll on mount/focus or 30-60s, never per message. Stale numbers acceptable.
4. Scope is TENANT balance. Schema has no per-user allowance; do not fake one off `request_attempts.user_id`.
5. Hive Enterprise posture: prepaid credits are a hosted-SaaS concept. Gate the widget on posture; Enterprise shows consumption against customer caps or nothing.

## Constraints

- Never expose `tenant_billing_accounts` to the browser.
- Do not add a stored balance column (#918 shows how derived-vs-stored shortcuts went wrong).
- All money math uses math/big; CREDITS_PER_USD = 100000.

### PR DESCRIPTION
**Pull request by sakibsadmanshajib**
## What

Closes #1063 (the surface half; the endpoint it asked for is included).

A signed-in chat user had no way to see credits or usage anywhere. This adds a small pill above the chat composer showing credits remaining and today's spend, with Claude-style phrasing, dismissible for the session.

## Data source

The balance is derived, never stored (`credit_ledger_entries` summed by `ledger.Service.GetBalance`, reused untouched). The gap #1063 identified was identity: chat sessions authenticate as Open WebUI users in `tenant_users`, a disjoint id space from `account_memberships`, so `/api/v1/accounts/current/credits/balance` is unusable from chat as-is.

New internal route on control-plane:

```
POST /internal/chat/credits/balance   {"email": "..."}
  -> 200 {posted_credits, reserved_credits, available_credits, usage_today_credits}
  -> 404 no billed account (banner hides)
  -> 400 empty/oversized body
```

Mounted behind `RequireInternalToken`, and only when a nonempty internal token is configured. The OWUI backend shim (owui-patches/hive_credits.py) resolves its own signed-in session to an email server side and calls this route with the `X-Internal-Token` header; the browser only ever sees the trimmed numbers under Open WebUI's own session. `tenant_billing_accounts` never reaches a browser.

Multi-tenant members resolve to the tenant selected in the product (`raw_user_meta_data->>'selected_tenant_id'`, written by /v1/tenants/switch), newest active membership as fallback; malformed selections degrade to the fallback order.

## UI

- `src/lib/hive/CreditsBanner.svelte`: three states.
  - healthy/low (below $0.50 in credits): "You've used N credits today · M remaining", amber at low.
  - empty: "You're out of credits." + Top up link when `HIVE_CONSOLE_BILLING_URL` is configured (deployment configuration served through the shim as `top_up_url`).
- Renders above both composer paths (conversation MessageInput and the landing Placeholder).
- Dismissible for the browsing session (sessionStorage, survives remounts).
- Any fetch failure degrades to no banner, zero console spam.
- MessageInput.svelte submit logic untouched.

## Enterprise posture

Prepaid credits are a hosted-SaaS concept. Deployments that do not carry `CONTROL_PLANE_INTERNAL_TOKEN` on the chat container get 404 and no banner (control-plane does not even mount the route); silent absence is the posture gate, documented in compose comments.

## Tests

- Go unit tests (stub repo): resolution, not-found semantics, handler contract incl. method/body/email edge cases, gate wiring.
- Go live-DB tests (HIVE_TEST_DB_URL-gated, run against the full migrated schema): four-table join, case-insensitive email, selected-tenant-wins, malformed-selection fallback. Verified locally against a migrated throwaway Postgres.
- Frontend vitest suite (81 tests green): state thresholds, dismissal persistence + storage-absence degradation, fetch success/non-200/network-failure silence, top_up_url passthrough.

## Visual proof

Posted below against a stack built from this branch (frontend + control-plane internal route, full migrated schema, signed-in chat user): healthy, low, and empty banner states. Comment: https://github.com/sakibsadmanshajib/hive/pull/1119#issuecomment-5400860095 ; capture log: docs/proof/usage-at-composer/capture-2026-08-24.md

## Review streams

- CodeRabbit GitHub app: ran (5 findings). All addressed or rebutted; every thread resolved.
- Plain adversarial pass: ran. Four findings fixed pre-push (dismissal persistence across remounts, payload coercion inside the failure path, server-side upstream-failure logging, i18n) plus one integration bug caught only by the live proof stack (shim sent the shared secret on `Authorization: Bearer`; RequireInternalToken reads `X-Internal-Token`): fixed and verified end to end.
- CodeRabbit CLI: SKIPPED, twice, WebSocket connection error from this environment. Noted per pipeline fallback rules.

## Buglog entry

Carried in commit 41d070f83 (wrong auth header between shim and RequireInternalToken, found by live-stack verification); to be appended to .wolf/buglog.jsonl on main via a buglog-only PR once this merges, per the issue #873 protocol.

### CODE REVIEW

File: apps/control-plane/cmd/server/main.go [hunk 1]
```diff
@@ -1363,6 +1363,17 @@ func main() {
...
 	log.Println("licensing entitlement route registered (issue #304)")
 
+	// Issue #1063: the chat composer's credits banner. Open WebUI's backend
+	// resolves its own signed-in user to an email server side and calls this
+	// internal route; the tenant->billing-account link stays behind the
+	// shared-secret gate and never reaches a browser.
+	if ledgerSvc != nil {
+		ledger.RegisterChatBalanceRoute(routerMux, ledgerSvc, func(h http.Handler) http.Handler {
+			return platformhttp.RequireInternalToken(cfg.InternalToken, h)
+		})
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/control-plane/cmd/server/main.go]: Fixed in b982e72b8.

Both halves of the ask, scoped to this change:

- `apps/control-plane/cmd/server/main.go` now mounts the chat-balance route only when a nonempty internal token is configured. RequireInternalToken already failed closed on an empty token, but skipping the mount entirely is the better failure shape: a misconfigured deployment reads as silent absence, not a live surface behind a well-known secret.
- The two compose lines added for the open-webui service pass `${CONTROL_PLANE_INTERNAL_TOKEN:-}` through with no committed fallback. With the token unset, control-plane does not mount the route and the shim answers 404, so no banner renders; that is also the Hive Enterprise posture gate.

Not changed here, deliberately: the pre-existing `hive-local-internal-token` default on the control-plane service's own env block (and edge-api's) predates this PR and is shared by other services' local boots. Removing it belongs in its own PR so it does not silently break unrelated profiles; happy to take that up separately.

File: apps/control-plane/internal/ledger/repository.go [hunk 2]
```diff
@@ -439,3 +445,56 @@ func isUniqueViolation(err error) bool {
...
+	err := r.pool.QueryRow(ctx, `
+		SELECT tba.account_id
+		FROM auth.users u
+		JOIN public.tenant_users tu ON tu.user_id = u.id AND tu.status = 'ACTIVE'
+		JOIN public.tenants t ON t.id = tu.tenant_id AND t.archived_at IS NULL
+		JOIN public.tenant_billing_accounts tba ON tba.tenant_id = t.id
+		WHERE lower(u.email) = lower($1)
+		ORDER BY tu.joined_at DESC
+		LIMIT 1
+	`, strings.TrimSpace(strings.ToLower(email))).Scan(&accountID)
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/control-plane/internal/ledger/repository.go]: Fixed in b982e72b8, with one deliberate deviation from the letter of the ask.

Resolution now prefers the tenant the user selected in the product: /v1/tenants/switch writes raw_user_meta_data->>'selected_tenant_id', and the query ranks a candidate membership whose tenant id matches that selection above everything else, with the newest-membership order as fallback. That is the trusted tenant identity this integration already has, without inventing a new wire field: the selection is server-owned state in auth.users, and it is exactly what 'the tenant this user chats under' means today. The comparison is text-vs-text so a malformed selection degrades to the fallback ordering (22P02 would turn a display surface into an error surface).

Regression test added at chat_balance_live_test.go (selected_tenant_wins_over_newer_membership): no-selection resolves newest, selected older tenant wins, malformed selection falls back. Passing against the real migrated schema.

Deviation: ambiguous memberships without any selection still resolve to the newest rather than suppressing the banner. Suppression there would punish the common case (a fresh signup whose newest tenant is their only billed tenant) for the rare one; with the selection preference in place the remaining ambiguity requires multi-tenant membership AND never having switched, and newest matches what signup provisions. If you want strict suppression anyway, say so and I will flip it.

File: apps/control-plane/internal/ledger/repository.go [hunk 3]
```diff
@@ -439,3 +445,56 @@ func isUniqueViolation(err error) bool {
...
+// an error, matching how GetBalance treats empty history.
+func (r *pgxRepository) GetUsageSince(ctx context.Context, accountID uuid.UUID, since time.Time) (int64, error) {
+	var used int64
+	err := r.pool.QueryRow(ctx, `
+		SELECT COALESCE(SUM(-credits_delta), 0)
+		FROM public.credit_ledger_entries
+		WHERE account_id = $1
+		  AND entry_type = 'usage_charge'
+		  AND created_at >= $2
+	`, accountID, since).Scan(&used)
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply apps/control-plane/internal/ledger/repository.go]: Rebutting: int64 is the correct contract here, not an oversight.

The ledger package's established representation for credit amounts is int64 end to end: BalanceSummary.PostedCredits / ReservedCredits / AvailableCredits (this same file, GetBalance), LedgerEntry.CreditsDelta, and the credits_delta bigint column underneath. GetUsageSince returns the same unit, summed with SQL SUM over that bigint column; there is no float anywhere on the path, so the rule math/big guards against (float corruption in money math) cannot occur. The math/big requirement bites at the FX conversion boundary, which this read-only display aggregate never approaches.

Converting one field of one response to big.Int would make it the only big.Int value in a package whose every consumer speaks int64, forcing conversion churn across handler, shim, and frontend for zero precision gain, and would diverge from the sibling fields it renders beside. If the package ever moves its whole contract to big.Int, this endpoint moves with it in that change.

File: vendor/open-webui/src/lib/components/chat/Chat.svelte [hunk 4]
```diff
@@ -3285,6 +3288,8 @@
 								</div>
 							{:else}
 								<div class=" pb-2 {dragged ? 'z-0' : 'z-10'}">
+									<!-- hive (#1063): credits pill above the composer -->
+									<CreditsBanner />
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply vendor/open-webui/src/lib/components/chat/Chat.svelte]: Fixed in b982e72b8. The banner now also renders above Placeholder on the landing path (no messages yet), inside a w-full wrapper so the existing flex row and Placeholder's own centering are untouched. Both composer paths carry it: the conversation view above MessageInput, the landing view above Placeholder.

The settings-declutter guard tests that pin Chat.svelte's source still pass (full hive frontend suite green at 81 tests).

File: vendor/open-webui/src/lib/hive/CreditsBanner.svelte [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+					? 'border-amber-200 bg-amber-50 text-amber-800 dark:border-amber-800 dark:bg-amber-950/40 dark:text-amber-300'
+				: 'border-gray-200/60 bg-gray-50 text-gray-500 dark:border-gray-850 dark:bg-gray-850/40 dark:text-gray-400'}"
+			role="status"
+		>
+			{#if state === 'empty'}
+				<span>{$i18n.t("You're out of credits.")}</span>
+				<a
+					href="https://console-hive.scubed.co/console/billing"
+					target="_blank"
+					rel="noreferrer"
```
[the remark this answers is not in the document]
Comment (sakibsadmanshajib) [reply vendor/open-webui/src/lib/hive/CreditsBanner.svelte]: Fixed in b982e72b8. The URL is deployment configuration now: the chat backend shim reads HIVE_CONSOLE_BILLING_URL from its environment and serves it as top_up_url on the balance payload; the banner renders the Top up link only when that value is configured and nonempty, and shows the out-of-credits warning either way. No production host in frontend source; compose passes the variable through with no default, so staging and self-hosted deployments point it wherever their console lives.

**Comment by sakibsadmanshajib** [comment]
## Visual proof

Credits banner above the composer, issue #1063. Healthy: used-today plus remaining. Low: below $0.50, amber. Empty: warning plus configured Top up link. Captured against a stack built from this branch (frontend + control-plane internal route), signed-in chat user, full migrated schema.
